### PR TITLE
MKAJ-4 add Builder

### PR DESCRIPTION
**Pull request by matsukata**
MKAJ-4 add Builder
- add Builder in Airplane and CrewMember entities
- refactor AirplaneDaoImpl
- refactor CrewMemberDaoImpl

### CODE REVIEW

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 1]
```diff
@@ -19,9 +19,9 @@
 public class AirplaneDaoImpl implements AirplaneDao {
     private static final String INSERT_QUERY = "INSERT INTO airplanes " +
             "(codeName, model, manufactureDate, capacity, flightRange, crew) VALUES (?, ?, ?, ?, ?, ?) ;";
-    private static final String FIND_ALL_SQL = "SELECT * FROM airplanes;";
-    private static final String SELECT_BY_CODE_NAME_SQL = "SELECT * FROM airplanes WHERE codeName = ?;";
-    private static final String DELETE_AIRPLANE_SQL = "delete from airplanes where id = ?;";
+    private static final String FIND_ALL_QUERY = "SELECT * FROM airplanes;";
+    private static final String SELECT_BY_CODE_NAME_QUERY = "SELECT * FROM airplanes WHERE codeName = ?;";
+    private static final String DELETE_QUERY = "delete from airplanes where id = ?;";
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: How this change is connected with your task?
As far as I can see, it should be discarded unless you can tell why it should be in that branch and what connection between task with implementing  builder pattern for entity and refactoring sql queries names in DaoImpl.

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 2]
```diff
@@ -33,29 +33,29 @@ public AirplaneDaoImpl(DataSource dataSource) {
     public void save(Airplane airplane) {
         try (Connection connection = dataSource.getConnection();
              PreparedStatement insertStatement = connection.prepareStatement(INSERT_QUERY,
-                     PreparedStatement.RETURN_GENERATED_KEYS);) {
+                     PreparedStatement.RETURN_GENERATED_KEYS)) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: You can leave this change, but for future again, don't forget that you can not just refactor and change any code you like in a branch, that sounds like "entity refactor"

All your changes in branch should be connected with your specific task you have.

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 3]
```diff
@@ -33,29 +33,29 @@ public AirplaneDaoImpl(DataSource dataSource) {
...
             insertStatement.executeUpdate();
             ResultSet generatedKeys = insertStatement.getGeneratedKeys();
             if (generatedKeys.next()) {
                 long id = generatedKeys.getLong(1);
-                airplane.setId(id);
+                Airplane.builder().withId(id);
             }
         } catch (SQLException e) {
-            throw new DaoOperationException("Error saving product: " + airplane, e);
+            throw new DaoOperationException("Error saving airplane: " + airplane, e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: How is this change related with your task? It's not entity.
Your branch should contain changes only for specific task, this change should be discarded

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 4]
```diff
@@ -33,29 +33,29 @@ public AirplaneDaoImpl(DataSource dataSource) {
...
         }
     }
 
     @Override
     public List<Airplane> findAll() {
         List<Airplane> airplaneArray = new ArrayList<>();
         try (Connection connection = dataSource.getConnection();
-             PreparedStatement preparedStatement = connection.prepareStatement(FIND_ALL_SQL)) {
+             PreparedStatement preparedStatement = connection.prepareStatement(FIND_ALL_QUERY)) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: This change is not part of your task, this change should be discarded

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 5]
```diff
@@ -64,50 +64,56 @@ public void save(Airplane airplane) {
...
+                        .withCodeName(codeName)
+                        .withModel(model)
+                        .withManufactureDate(manufactureDate)
+                        .withCapacity(capacity)
+                        .withFlightRange(flightRange)
+                        .withCrewMember(crewMember).build());
             }
         } catch (SQLException e) {
-            throw new DaoOperationException("Couldn't save product ", e);
+            throw new DaoOperationException("Couldn't find airplane", e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: This change should be discarded because it's out of scope of your task

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 6]
```diff
@@ -64,50 +64,56 @@ public void save(Airplane airplane) {
...
+                        .withModel(resultSet.getString("model"))
+                        .withManufactureDate(resultSet.getDate("manufactureDate").toLocalDate())
+                        .withCapacity(resultSet.getInt("capacity"))
+                        .withFlightRange(resultSet.getInt("flightRange"))
+                        .build();
             }
-        } catch (SQLException throwables) {
-            throwables.printStackTrace();
+        } catch (SQLException e) {
+            throw new DaoOperationException("Could not find airplane", e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: This change should be discarded, because it's out of scope of your task

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 7]
```diff
@@ -64,50 +64,56 @@ public void save(Airplane airplane) {
...
         }
         try (Connection connection = dataSource.getConnection();
-             PreparedStatement statement = connection.prepareStatement(DELETE_AIRPLANE_SQL);) {
+             PreparedStatement statement = connection.prepareStatement(DELETE_QUERY)) {
             statement.setLong(1, airplane.getId());
             statement.executeUpdate();
-        } catch (SQLException throwables) {
-            throwables.printStackTrace();
+        } catch (SQLException e) {
+            throw new DaoOperationException("Could not remove airplane", e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: This change should be discarded, because it's out of scope of your task

File: src/main/java/com/airlines/model/CrewMember.java [hunk 8]
```diff
@@ -54,60 +40,21 @@ public LocalDate getBirthday() {
...
-            return false;
-        }
-        if (position != null ? !position.equals(crew.position) : crew.position != null) {
-            return false;
-        }
-        if (birthday != null ? !birthday.equals(crew.birthday) : crew.birthday != null) {
-            return false;
-        }
-        return citizenship != null ? citizenship.equals(crew.citizenship) : crew.citizenship == null;
+        if (this == o) return true;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/CrewMember.java]: curly braces are missing

File: src/main/java/com/airlines/model/CrewMember.java [hunk 9]
```diff
@@ -1,51 +1,37 @@
...
-        this.birthday = birthday;
-        this.citizenship = citizenship;
+    private final Long id;
+    private final String firstName;
+    private final String lastName;
+    private final Position position;
+    private final LocalDate birthday;
+    private final Citizenship citizenship;
+
+    public CrewMember(Builder builder) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/CrewMember.java]: for what purpose CrewMember has public access, where it could be used?

File: src/main/java/com/airlines/model/Crew.java [hunk 10]
```diff
@@ -20,44 +18,25 @@ public Long getId() {
...
         return crewMembers;
     }
 
-    public void setCrewMembers(List<CrewMember> crewMembers) {
-        this.crewMembers = crewMembers;
-    }
-
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Crew.java]: curly braces are missing

File: src/main/java/com/airlines/model/Crew.java [hunk 11]
```diff
@@ -1,16 +1,14 @@
...
+    private final List<CrewMember> crewMembers;
 
-    public Crew() {
-    }
-
-    public Crew(Long id, String name, List<CrewMember> crewMembers) {
+    Crew(Long id, String name, List<CrewMember> crewMembers) {
         this.id = id;
         this.name = name;
         this.crewMembers = crewMembers;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Crew.java]: id and name are immutable, we can't change them after Crew instance is created.
But List<CrewMember> could be changed, because final modifier means we are obligated to assign reference to some List object to our variable - clas field crewMembers, and we can not reattach other list.

But we can change state of this list:
Crew crew = new Crew(1L, "Blue Team", crewMembers);
List<CrewMember> list = cerw.getCrewMembers();
list.clear();
list.add(badCrewMember);

and so on.

So as long as we can get link to mutable object, we can change state, so Crew is mutable, we can change it any time.

I don't know if we need immutable there, but to get known with this moment I suggest making field "crewMembers" immutable

you can do it in constructor.

Google
"how make list immutable java" there special Collections class method

File: src/main/java/com/airlines/model/Airplane.java [hunk 12]
```diff
@@ -118,7 +72,63 @@ public String toString() {
                 ", manufactureDate=" + manufactureDate +
                 ", capacity=" + capacity +
                 ", flightRange=" + flightRange +
-                ", crew=" + crew +
+                ", crew=" + crewMember +
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Airplane.java]: why do you think it should be crewMember? 
I don't  see crewMember in Airplanes database scheme

File: src/main/java/com/airlines/model/Airplane.java [hunk 13]
```diff
@@ -1,112 +1,66 @@
...
-    public void setCrew(CrewMember crew) {
-        this.crew = crew;
+    public CrewMember getCrewMember() {
+        return crewMember;
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Airplane.java]: curly braces are missing here

File: src/main/java/com/airlines/model/Airplane.java [hunk 14]
```diff
@@ -1,112 +1,66 @@
...
-        this.capacity = capacity;
-        this.flightRange = flightRange;
-        this.crew = crew;
+    private final Long id;
+    private final String codeName;
+    private final String model;
+    private final LocalDate manufactureDate;
+    private final int capacity;
+    private final int flightRange;
+    private final CrewMember crewMember;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Airplane.java]: why you think Airplane has only one CrewMember? 
Your entity doesn't match to SQL database scheme

File: src/main/java/com/airlines/model/Airplane.java [hunk 15]
```diff
@@ -1,112 +1,66 @@
...
+
+    public Airplane(Builder builder) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/model/Airplane.java]: why you need this constructor as public?

File: src/main/java/com/airlines/dao/impl/CrewMemberDaoImpl.java [hunk 16]
```diff
@@ -50,19 +50,22 @@ public void save(CrewMember crewMember) {
...
-                crewMember.setPosition(Position.valueOf(resultSet.getString("position")));
-                crewMember.setCitizenship(Citizenship.valueOf(resultSet.getString("citizenship")));
+                crewMember = CrewMember.builder()
+                        .withId(resultSet.getLong("id"))
+                        .withFirstName(resultSet.getString("first_name"))
+                        .withLastName(resultSet.getString("last_name"))
+                        .withPosition(Position.valueOf(resultSet.getString("position")))
+                        .withBirthday(resultSet.getDate("birthday").toLocalDate())
+                        .withCitizenship(Citizenship.valueOf(resultSet.getString("citizenship")))
+                        .build();
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/CrewMemberDaoImpl.java]: I guess you could do next:
1) remove  `return crewMember;` line from the end of method
2) remove line 55` rewMember crewMember = null;`
3) 61 line change: from
`crewMember = CrewMember.builder()   `
-> to
`return  CrewMember.builder()...`

File: src/main/java/com/airlines/dao/impl/CrewMemberDaoImpl.java [hunk 17]
```diff
@@ -40,7 +40,7 @@ public void save(CrewMember crewMember) {
             ResultSet generatedKeys = insertStatement.getGeneratedKeys();
             if (generatedKeys.next()) {
                 long id = generatedKeys.getLong(1);
-                crewMember.setId(id);
+                CrewMember.builder().withId(id);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/CrewMemberDaoImpl.java]: This line doesn't have any sense I guess.

Previous flow with crewMember.setId(id);

we have some object, CrewMember with all fields but without id. We save it to database. And set id for this entity.
So:
```
CrewMember draft = new CraftMember(...) //without id
save(draft);  //saving draft to database and setting id to this object
...
Long id = draft.getId(); // now we have our id a
```

But with builder it became more complicated. Our  entity is immutable. It couldn't be updated with id.
And this line:
` CrewMember.builder().withId(id);`
Means that new CrewMember object is created, with only one field - id.
All other fields are missing , you could easily check it by making 
public static void main (String...args) method
 and code there:
CrewMember crewMember = CrewMember.builder().withId(id);
System.out.println(cerwMember);
only id is set.

and moreover, 
line
` CrewMember.builder().withId(id);`
do nothing, because this line returns builder, and you did nothing with it.

It's like:

`list.get(0);`
line, you retrieve element, but you do nothing with this object, Usually, idea will mark it by grey or yellow color and message:
"return value is ignored"

I don't know actually, what it's better to do...I guess we have two options:
1) just drop 
```
        if (generatedKeys.next()) {
                long id = generatedKeys.getLong(1);
                CrewMember.builder().withId(id);
            }
```
because it do nothing

2)  change method signature to:
`public CrewMember save(CrewMember crewMember) {`
then
```
        if (generatedKeys.next()) {
                long id = generatedKeys.getLong(1);
                return copyAttributesAndSetId(crewMember, id);
            }
```
 private CrewMember copyAttributesAndSetId(CrewMember crewMember, Long id) {
   return CrewMember.builder()
            .withId(id)
           .withFirstName(crewMember.getFirstName)
           ...
}
and just copy all attributes.

Do option #2

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 18]
```diff
@@ -64,50 +64,56 @@ public void save(Airplane airplane) {
...
         if (airplane.getId() == null) {
-            throw new DaoOperationException("Cannot find a product without ID");
-        }
-        if (airplane.getId() < 0) {
-            throw new DaoOperationException(String.format("Product with id = %d does not exist", airplane.getId()));
+            throw new DaoOperationException("Cannot remove an airplane without id");
         }
         try (Connection connection = dataSource.getConnection();
-             PreparedStatement statement = connection.prepareStatement(DELETE_AIRPLANE_SQL);) {
+             PreparedStatement statement = connection.prepareStatement(DELETE_QUERY)) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: change should be discarded as out of scope

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 19]
```diff
@@ -64,50 +64,56 @@ public void save(Airplane airplane) {
...
     }
 
     @Override
     public void remove(Airplane airplane) {
         if (airplane.getId() == null) {
-            throw new DaoOperationException("Cannot find a product without ID");
-        }
-        if (airplane.getId() < 0) {
-            throw new DaoOperationException(String.format("Product with id = %d does not exist", airplane.getId()));
+            throw new DaoOperationException("Cannot remove an airplane without id");
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: should be discarded because out of scope

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 20]
```diff
@@ -33,29 +33,29 @@ public AirplaneDaoImpl(DataSource dataSource) {
...
             insertStatement.setInt(4, airplane.getCapacity());
             insertStatement.setInt(5, airplane.getFlightRange());
-            insertStatement.setObject(6, airplane.getCrew());
+            insertStatement.setObject(6, airplane.getCrewMember());
             insertStatement.executeUpdate();
             ResultSet generatedKeys = insertStatement.getGeneratedKeys();
             if (generatedKeys.next()) {
                 long id = generatedKeys.getLong(1);
-                airplane.setId(id);
+                Airplane.builder().withId(id);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: Do same changes as for CrewDaoImpl : (copy attributes and set id, change signature)

**Review by joseph-dredd — CHANGES_REQUESTED** [review]

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 21]
```diff
@@ -44,13 +44,27 @@ public void save(Airplane airplane) {
...
             if (generatedKeys.next()) {
                 long id = generatedKeys.getLong(1);
-                airplane.setId(id);
+                return Optional.of(copyAttributesAndSetAirplaneId(airplane, id));
+            } else {
+                return Optional.empty();
             }
         } catch (SQLException e) {
-            throw new DaoOperationException("Error saving product: " + airplane, e);
+            throw new DaoOperationException("Couldn't save product ", e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: How this change connect with task "Entity upgrade with builder"?
Discard this change

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 22]
```diff
@@ -75,18 +97,21 @@ public void save(Airplane airplane) {
...
     public Airplane findOne(String codeName) {
-        Airplane airplane = new Airplane();
+        Airplane airplane = null;
         try (Connection connection = dataSource.getConnection();
-             PreparedStatement selectByIdStatement = connection.prepareStatement(SELECT_BY_CODE_NAME_SQL);) {
-            selectByIdStatement.setString(1, codeName);
-            ResultSet resultSet = selectByIdStatement.executeQuery();
+             PreparedStatement selectByCodeNameStatement = connection.prepareStatement(SELECT_BY_CODE_NAME_SQL);) {
+            selectByCodeNameStatement.setString(1, codeName);
+            ResultSet resultSet = selectByCodeNameStatement.executeQuery();
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: How this change connected with your task?
You could change this DaoImpl only in part where you change old entity style (constructor + setters) to new one (builder).
Refactoring whole project is not part of story
Discard this change

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 23]
```diff
@@ -103,7 +128,7 @@ public void remove(Airplane airplane) {
             throw new DaoOperationException(String.format("Product with id = %d does not exist", airplane.getId()));
         }
         try (Connection connection = dataSource.getConnection();
-             PreparedStatement statement = connection.prepareStatement(DELETE_AIRPLANE_SQL);) {
+             PreparedStatement statement = connection.prepareStatement(DELETE_AIRPLANE_SQL)) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: DaoImpl is not part of your task
Discard this change

File: src/main/java/com/airlines/exception/DaoOperationException.java [hunk 24]
```diff
@@ -9,4 +9,3 @@ public DaoOperationException(String message, Throwable cause) {
         super(message, cause);
     }
 }
-
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/exception/DaoOperationException.java]: discard this change

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java [hunk 25]
```diff
@@ -44,13 +44,27 @@ public void save(Airplane airplane) {
...
+                .withId(id)
+                .withCodeName(airplane.getCodeName())
+                .withModel(airplane.getModel())
+                .withManufactureDate(airplane.getManufactureDate())
+                .withCapacity(airplane.getCapacity())
+                .withFlightRange(airplane.getFlightRange())
+                .withCrew(airplane.getCrew())
+                .build();
+    }
+
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: private methods always located after public methods

**Review by joseph-dredd — CHANGES_REQUESTED** [review]

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java (hunk 22 above)
Comment (matsukata) [reply src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: done in https://github.com/Matsukata/airlines-jdbc-app/pull/5/commits/625cab1de0c97028d188cc5a80294e281f46b375

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java (hunk 23 above)
Comment (matsukata) [reply src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: done in https://github.com/Matsukata/airlines-jdbc-app/pull/5/commits/625cab1de0c97028d188cc5a80294e281f46b375

File: src/main/java/com/airlines/exception/DaoOperationException.java (hunk 24 above)
Comment (matsukata) [reply src/main/java/com/airlines/exception/DaoOperationException.java]: done in https://github.com/Matsukata/airlines-jdbc-app/pull/5/commits/625cab1de0c97028d188cc5a80294e281f46b375

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java (hunk 25 above)
Comment (matsukata) [reply src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: done in https://github.com/Matsukata/airlines-jdbc-app/pull/5/commits/625cab1de0c97028d188cc5a80294e281f46b375

File: src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java (hunk 21 above)
Comment (matsukata) [reply src/main/java/com/airlines/dao/impl/AirplaneDaoImpl.java]: done in https://github.com/Matsukata/airlines-jdbc-app/pull/5/commits/625cab1de0c97028d188cc5a80294e281f46b375

**Review by joseph-dredd — APPROVED** [review]
